support/cas-server-support-fortress/src/main/java/org/apereo/cas/adaptors/fortress/FortressAuthenticationHandler.java

```diff
@@ -0,0 +1,97 @@
+    private static final Logger LOG = LoggerFactory.getLogger(FortressAuthenticationHandler.class);
+    private static final String FORTRESS_SESSION_KEY = "fortressSession";
+    private static final String JAXB_EXCEPTION_MESSAGE = "failed initializa jaxb context";
+    private static final String TRACE_ATTEMPT_MESSAGE_TO_FORTRESS = "trying to delegate authentication for {} to fortress";
+    private static final String TRACE_FORTRESS_SESSION_RESPONSE = "fortress session result : {}";
+    private static final String TRACE_FORTRESS_AUTH_SUCCESS = "returning default handler";
+
+    @Autowired
+    @Qualifier("fortressAccessManager")
+    private AccessMgr accessManager;
+
+    private Marshaller marshaller;
+
+    public FortressAuthenticationHandler(final String name, final ServicesManager servicesManager,
+                                         final PrincipalFactory principalFactory, final Integer order) {
+        super(name, servicesManager, principalFactory, order);
+        try {
+            final JAXBContext jaxbContext = JAXBContext.newInstance(Session.class);
+            marshaller = jaxbContext.createMarshaller();
+        } catch (final JAXBException e) {
+            LOG.error(JAXB_EXCEPTION_MESSAGE, e);
+        }
+        LOG.trace("Fortress authentication handler registered");
+    }
+
+    @Override
+    protected HandlerResult authenticateUsernamePasswordInternal(final UsernamePasswordCredential usernamePasswordCredential,
+                                                                 final String originalPassword) throws GeneralSecurityException, PreventedException {
+        final String username = usernamePasswordCredential.getUsername();
+        final String password = usernamePasswordCredential.getPassword();
+        Session fortressSession = null;
+        try {
+            LOG.trace(TRACE_ATTEMPT_MESSAGE_TO_FORTRESS, new Object[]{username});
+            fortressSession = accessManager.createSession(new User(username, password.toCharArray()), false);
+            if (fortressSession != null) {
+                final StringWriter writer = new StringWriter();
+                marshaller.marshal(fortressSession, writer);
+                final String fortressXmlSession = writer.toString();
+                LOG.trace(TRACE_FORTRESS_SESSION_RESPONSE, fortressXmlSession);
+                final Map<String, Object> attributes = new HashMap<>();
+                attributes.put(FORTRESS_SESSION_KEY, fortressXmlSession);
+                return createHandlerResult(usernamePasswordCredential,
+                        principalFactory.createPrincipal(username, attributes), null);
+            }
+        } catch (final org.apache.directory.fortress.core.SecurityException e) {
+            final String errorMessage = "Fortress authentication failed for [" + username + "]";
+            LOG.trace(errorMessage, e);
+            throw new GeneralSecurityException(errorMessage);
```

Comment:
Note that this is probably not the best exception to throw. There are better ones that specifically deal with authn failures. See other handlers for examples.